File: src/main/java/org/example/expert/domain/todo/repository/TodoRepositoryCustom.java
package org.example.expert.domain.todo.repository;

import java.time.LocalDate;
import java.util.Optional;
import org.example.expert.domain.todo.dto.response.TodoSearchResponse;
import org.example.expert.domain.todo.entity.Todo;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface TodoRepositoryCustom {

    Optional<Todo> findByIdWithUser(Long todoId);

    Page<TodoSearchResponse> searchWithRelationCounts(
        String keyword,
        LocalDate startDate,
        LocalDate endDate,
        String managerName,
        Pageable pageable
    );
}

File: src/main/kotlin/org/example/expert/domain/todo/repository/TodoRepositoryCustom.kt
package org.example.expert.domain.todo.repository

import org.example.expert.domain.todo.dto.response.TodoSearchResponse
import org.example.expert.domain.todo.entity.Todo
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import java.time.LocalDate

interface TodoRepositoryCustom {

    fun findByIdWithUser(todoId: Long): Todo?

    fun searchWithRelationCounts(
        keyword: String?,
        startDate: LocalDate?,
        endDate: LocalDate?,
        managerName: String?,
        pageable: Pageable,
    ): Page<TodoSearchResponse>
}